src/mds/SessionMap.cc

```diff
@@ -449,3 +656,41 @@ void Session::decode(bufferlist::iterator &p)
   _update_human_name();
 }
 
+/**
+ * Mark a session as requiring write by the current value of `version`.
+ */
+void SessionMap::mark_dirty(const Session *s)
+{
+  dout(20) << __func__ << " s=" << s << " name=" << s->info.inst.name << dendl;
+  dirty_sessions.insert(s->info.inst.name);
+}
+
+/**
+ * Start a new version: this will make the 
+ */
+void SessionMap::inc_version()
+{
+  dout(20) << __func__ << " v " << version << " -> " << version + 1
+    << " dirty=" << dirty_sessions.size() << "/"
+    << g_conf->mds_sessionmap_keys_per_op / 2 << dendl;
+
+  if (dirty_sessions.size() >= g_conf->mds_sessionmap_keys_per_op / 2) {
+    // If there are >= allowance/2 sessions dirty, then we have to
+    // do a save now, because otherwise the sessions written in this
+    // newly incremented version could leave us with more than `allowance`
+    // dirty sessions, and thereby an oversized write in the next save()
+    save(new C_MDSInternalNoop, version);
```

Comment:
Oh right, we have to handle projected versions for the logging. Still, I think we can do that (and make the privacy better) by introducing some kind of "project_dirty_session()" that returns the projected version it will use? And then we can guarantee (with some checks via a registry) that all projected dirty sessions are marked as dirty in order, and so they get the right versions attached?